ably.ts.d

```diff
@@ -0,0 +1,400 @@
+
+// Interfaces
+
+interface ClientOptions extends AuthOptions {
+  autoConnect?: boolean;
+  clientId?: String;
+  defaultTokenParams?: TokenParams;
+  echoMessages?: boolean;
+  environment?: String;
+  log?: LogInfo;
+  port?: number;
+  queueMessages?: boolean;
+  restHost?: String;
+  realtimeHost?: String;
+  fallbackHosts?: Array<String>;
+  recover?: String | standardCallback;
+  tls?: boolean;
+  tlsPort?: number;
+  useBinaryProtocol?: boolean;
+}
+
+interface AuthOptions {
+  authCallback?: (data: TokenParams,callback: (Error: ErrorInfo | String, tokenRequestOrDetails: String | TokenDetails | TokenRequest) => void) => void;
+  authHeaders?: Array<any>;
+  authMethod?: String;
+  authParams?: Array<any>;
+  authUrl?: String;
+  key?: String;
+  queryTime?: boolean;
+  token?: String | TokenDetails;
+  tokenDetails?: TokenDetails;
+  useTokenAuth?: boolean;
+}
+
+interface TokenParams {
+  capability?: String;
+  clientId?: String;
+  nonce?: String;
+  timestamp?: Number;   // Time?
+  ttl?: Number;
+}
+
+interface CipherParams {
+  algorithm: String;
+  key: any;
+  keyLength: number;
+  mode: String;
+}
+
+interface ErrorInfo {
+  code: number;
+  message: String;
+  statusCode: number;
+}
+
+interface StatsMessageTypes {
+  all: number,
+  messages: number,
+  presence: number;
+}
+
+interface StatsRequestCount {
+  failed: number,
+  refused: number,
+  succeeded: number
+}
+
+interface StatsResourceCount {
+  mean: number,
+  min: number,
+  opened: number,
+  peak: number,
+  refused: number
+}
+
+interface StatsConnectionTypes {
+  all: number,
+  plain: number,
+  tls: number
+}
+
+interface StatsMessageTraffic {
+  all: number,
+  realtime: number,
+  rest: number,
+  webhook: number
+}
+
+interface TokenDetails {
+  capability: String;
+  clientId: String;
+  expires: number; // Time
+  issued: number; // Time
+  token: String;
+}
+
+interface TokenRequest {
+  capability: String;
+  clientId?: String;
+  keyName: String;
+  mac: String;
+  nonce: String;
+  timestamp: number; // Time
+  ttl?: number;
+}
+
+interface ChannelOptions {
+  cipher: any;
+}
+
+interface RestPresenceHistoryParams {
+  start?: number;
+  end?: number;
+  direction?: String;
+  limit?: number;
+}
+
+
+interface RestPresenceParams {
+  limit?: number;
+  clientId?: String;
+  connectionId?: String;
+}
+
+interface RealtimePresenceParams {
+  waitForSync?: boolean;
+  clientId?: String;
+  connectionId?: String;
+}
+
+interface RealtimePresenceHistoryParams {
+  start?: number;
+  end?: number;
+  direction?: String;
+  limit?: number;
+  untilAttach?: boolean
+}
+
+interface LogInfo {
+  level?: number;
+  handler?: (...args) => void;
+}
+
+
+// Common Listeners
+type ListenerFn = (...args: Array<any>) => void;
+type PaginatedResultCallback = (Error: ErrorInfo, results: PaginatedResult ) => void;
+type standardCallback = (Error: ErrorInfo, results: any) => void;
+type messageCallback = (Msg: any) => void;
```

Comment:
Why not `Message` type?  Then perhaps add a `presenceMessageCallback` too?  Alternatively, perhaps better, create a new `subscribeCallback` using a generic that supports any type and specify the type later.